source/CODE.md

```diff
@@ -1,33 +1,377 @@
-  intermediate verification language based on assert and assume statements,
-  along with mutable updates to local variables.
-
-Additional files are found in:
-- `rust_verify_test_macros`: Provides useful macros for creating unit tests for
-  Verus.  Factored into a separate crate, due to the use of procedural macros.
-- `builtin`: Defines verification-related Rust "functions" (e.g., `assert`,
-  `decreases`), so that the Rust compiler will automatically create correctly
-  typed internal representations.  These are handled in a custom manner inside
-  of Verus.
-- `builtin_macros`: Provides useful macros for the `builtin` library.  These
-  are in a separate crate, since Rust requires procedural macros to be in their
-  own crate.
-- `tools`: Scripts for running Verus in various configurations.
+# Verus Architecture Overview
+
+This document provides an overview of Verus internals for developers
+contributing to the project. For build instructions and testing, see
+[CONTRIBUTING.md](../CONTRIBUTING.md).
+
+## Verification Pipeline Overview
+
+Verus transforms Rust source code through multiple intermediate representations:
+
+```
+Rust Source Code
+    ↓ (rustc: parsing, macro expansion)
+Rust HIR (High-level Intermediate Representation)
+    ↓ (rust_verify: HIR → VIR)
+VIR-AST (Verification Intermediate Representation)
+    ↓ (vir: AST → SST)
+VIR-SST (Statement-oriented Syntax Tree)
+    ↓ (vir: SST → AIR)
+AIR (Assertion Intermediate Representation)
+    ↓ (air: SMT encoding)
+SMT-LIB queries
+    ↓ (Z3, and experimentally cvc5)
+Verification Results
+```
+
+## Crate Architecture
+
+### Pipeline Call Graph Overview
+
+```
+main() → Verifier::new() → driver::run()
+    ↓
+VerifierCallbacksEraseMacro::after_expansion()
+    ↓
+Verifier::construct_vir_crate()
+    ├─ rustc_hir_analysis::check_crate()  [type check]
+    └─ rust_to_vir::crate_to_vir()        [HIR → VIR]
+    ↓
+Verifier::verify_crate_inner()
+    ├─ GlobalCtx::new()
+    ├─ ast_simplify::simplify_krate()
+    ├─ modes::check_crate()
+    ├─ buckets::get_buckets()
+    └─ [per bucket, in parallel]
+        └─ verify_bucket()
+            └─ [per function]
+                ├─ ast_to_sst_func::ast_to_sst_func()
+                ├─ sst_to_air_func::func_sst_to_air()
+                └─ Context::query() → SmtProcess::send()
+```
+
+
+### rust_verify
+
+The main verifier driver that integrates with `rustc`.  This is the only
+crate that should interact with `rustc`.  All other crates use Verus-specific
+constructs.
+
+**Key files:**
+- `main.rs`: Entry point, argument parsing, output formatting
+- `driver.rs`: Orchestrates `rustc` invocations for verification and compilation
+- `verifier.rs`: Core `Verifier` struct and verification logic
+- `context.rs`: `Context<'tcx>` carrying `rustc` state
+- `config.rs`: Command-line argument definitions (`Args` struct)
+- `attributes.rs`: Verus attribute parsing (`#[verifier(...)]`)
+- `lifetime.rs`: Lifetime and borrow checking integration
```

Comment:
these files need to be renamed. They used to do lifetime-checking + the trait conflict checker on the side. Then we rearchitected the lifetime system and moved it else. So now these files only do trait conflict checking

Reply:
Or perhaps you just meant the `lifetime_*.rs` files?

Reply:
Sorry to make your documentation life difficult >_< I left this file in a bad shape that doesn't make any sense.

Reply:
Okay, I moved `check` and the `rustc_driver` callbacks that invoke it into `driver.rs` (814e6ccac5d86b03) and then renamed the `lifetime*.rs` files to `trait_check*.rs` (56b8fb162c).  There are still function and constant names that refer to lifetimes, but I thought I'd leave that for a separate PR.